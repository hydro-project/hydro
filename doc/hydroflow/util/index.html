<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Helper utilities for the Hydroflow surface syntax."><title>hydroflow::util - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-fecee23daa4e8a6d.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="hydroflow" data-themes="" data-resource-suffix="" data-rustdoc-version="1.83.0-nightly (4ac7bcbaa 2024-09-04)" data-channel="nightly" data-search-js="search-a99f1315e7cc5121.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../static.files/storage-118b08c4c78b968e.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-921df33f47b8780c.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-3b12f09e550e0385.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../hydroflow/index.html">hydroflow</a><span class="version">0.9.0</span></h2></div><h2 class="location"><a href="#">Module util</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#functions">Functions</a></li><li><a href="#types">Type Aliases</a></li></ul></section><h2><a href="../index.html">In crate hydroflow</a></h2></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../index.html">hydroflow</a>::<wbr><a class="mod" href="#">util</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../../src/hydroflow/util/mod.rs.html#1-376">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Helper utilities for the Hydroflow surface syntax.</p>
</div></details><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="clear/index.html" title="mod hydroflow::util::clear">clear</a></div><div class="desc docblock-short">Module for the <a href="clear/trait.Clear.html" title="trait hydroflow::util::clear::Clear"><code>Clear</code></a> trait.</div></li><li><div class="item-name"><a class="mod" href="demux_enum/index.html" title="mod hydroflow::util::demux_enum">demux_<wbr>enum</a></div><div class="desc docblock-short">Trait for the <code>demux_enum</code> derive and operator.</div></li><li><div class="item-name"><a class="mod" href="deploy/index.html" title="mod hydroflow::util::deploy">deploy</a></div></li><li><div class="item-name"><a class="mod" href="monotonic_map/index.html" title="mod hydroflow::util::monotonic_map">monotonic_<wbr>map</a></div><div class="desc docblock-short">Module for <a href="monotonic_map/struct.MonotonicMap.html" title="struct hydroflow::util::monotonic_map::MonotonicMap"><code>MonotonicMap</code></a>.</div></li><li><div class="item-name"><a class="mod" href="multiset/index.html" title="mod hydroflow::util::multiset">multiset</a></div><div class="desc docblock-short">A multiset backed by a HashMap</div></li><li><div class="item-name"><a class="mod" href="simulation/index.html" title="mod hydroflow::util::simulation">simulation</a></div><div class="desc docblock-short">Hydroflow Deterministic Simulation Testing Framework</div></li><li><div class="item-name"><a class="mod" href="sparse_vec/index.html" title="mod hydroflow::util::sparse_vec">sparse_<wbr>vec</a></div><div class="desc docblock-short">A vector that supports efficient deletion without reordering all subsequent items.</div></li><li><div class="item-name"><a class="mod" href="unsync/index.html" title="mod hydroflow::util::unsync">unsync</a></div><div class="desc docblock-short">Unsync (i.e. single-threaded) synchronization primitives.</div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.DroppableChild.html" title="struct hydroflow::util::DroppableChild">Droppable<wbr>Child</a></div><div class="desc docblock-short">Terminates the inner <a href="https://doc.rust-lang.org/nightly/std/process/struct.Child.html" title="struct std::process::Child"><code>Child</code></a> process when dropped.</div></li><li><div class="item-name"><a class="struct" href="struct.MonotonicFn.html" title="struct hydroflow::util::MonotonicFn">Monotonic<wbr>Fn</a></div><div class="desc docblock-short">A wrapper christening a closure as a <a href="https://hydro.run/docs/hydroflow/lattices_crate/lattice_math#the-calm-theorem-and-monotonicity">monotonic function</a></div></li><li><div class="item-name"><a class="struct" href="struct.Morphism.html" title="struct hydroflow::util::Morphism">Morphism</a></div><div class="desc docblock-short">A wrapper christening a closure as a <a href="https://hydro.run/docs/hydroflow/lattices_crate/lattice_math#lattice-morphism">lattice morphism</a></div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.Persistence.html" title="enum hydroflow::util::Persistence">Persistence</a></div><div class="desc docblock-short">Persit or delete tuples</div></li><li><div class="item-name"><a class="enum" href="enum.PersistenceKeyed.html" title="enum hydroflow::util::PersistenceKeyed">Persistence<wbr>Keyed</a></div><div class="desc docblock-short">Persit or delete key-value pairs</div></li></ul><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.bind_tcp.html" title="fn hydroflow::util::bind_tcp">bind_<wbr>tcp</a></div><div class="desc docblock-short">Create a listening tcp socket, and then as new connections come in, receive their data and forward it to a queue.</div></li><li><div class="item-name"><a class="fn" href="fn.bind_tcp_bytes.html" title="fn hydroflow::util::bind_tcp_bytes">bind_<wbr>tcp_<wbr>bytes</a></div><div class="desc docblock-short">Returns a newline-delimited bytes <code>Sender</code>, <code>Receiver</code>, and <code>SocketAddr</code> bound to the given address.</div></li><li><div class="item-name"><a class="fn" href="fn.bind_tcp_lines.html" title="fn hydroflow::util::bind_tcp_lines">bind_<wbr>tcp_<wbr>lines</a></div><div class="desc docblock-short">This is the same thing as <code>bind_tcp_bytes</code> except instead of using a length-delimited encoding scheme it uses new lines to separate frames.</div></li><li><div class="item-name"><a class="fn" href="fn.bind_udp_bytes.html" title="fn hydroflow::util::bind_udp_bytes">bind_<wbr>udp_<wbr>bytes</a></div><div class="desc docblock-short">Returns a length-delimited bytes <code>Sink</code>, <code>Stream</code>, and <code>SocketAddr</code> bound to the given address.
The input <code>addr</code> may have a port of <code>0</code>, the returned <code>SocketAddr</code> will have the chosen port.</div></li><li><div class="item-name"><a class="fn" href="fn.bind_udp_lines.html" title="fn hydroflow::util::bind_udp_lines">bind_<wbr>udp_<wbr>lines</a></div><div class="desc docblock-short">Returns a newline-delimited bytes <code>Sink</code>, <code>Stream</code>, and <code>SocketAddr</code> bound to the given address.
The input <code>addr</code> may have a port of <code>0</code>, the returned <code>SocketAddr</code> will have the chosen port.</div></li><li><div class="item-name"><a class="fn" href="fn.collect_ready.html" title="fn hydroflow::util::collect_ready">collect_<wbr>ready</a></div><div class="desc docblock-short">Collects the immediately available items from the <code>Stream</code> into a <code>FromIterator</code> collection.</div></li><li><div class="item-name"><a class="fn" href="fn.collect_ready_async.html" title="fn hydroflow::util::collect_ready_async">collect_<wbr>ready_<wbr>async</a></div><div class="desc docblock-short">Collects the immediately available items from the <code>Stream</code> into a collection (<code>Default</code> + <code>Extend</code>).</div></li><li><div class="item-name"><a class="fn" href="fn.connect_tcp.html" title="fn hydroflow::util::connect_tcp">connect_<wbr>tcp</a></div><div class="desc docblock-short">The inverse of <a href="fn.bind_tcp.html" title="fn hydroflow::util::bind_tcp"><code>bind_tcp</code></a>.</div></li><li><div class="item-name"><a class="fn" href="fn.connect_tcp_bytes.html" title="fn hydroflow::util::connect_tcp_bytes">connect_<wbr>tcp_<wbr>bytes</a></div><div class="desc docblock-short">The inverse of <a href="fn.bind_tcp_bytes.html" title="fn hydroflow::util::bind_tcp_bytes"><code>bind_tcp_bytes</code></a>.</div></li><li><div class="item-name"><a class="fn" href="fn.connect_tcp_lines.html" title="fn hydroflow::util::connect_tcp_lines">connect_<wbr>tcp_<wbr>lines</a></div><div class="desc docblock-short">This is the same thing as <code>connect_tcp_bytes</code> except instead of using a length-delimited encoding scheme it uses new lines to separate frames.</div></li><li><div class="item-name"><a class="fn" href="fn.deserialize_from_bytes.html" title="fn hydroflow::util::deserialize_from_bytes">deserialize_<wbr>from_<wbr>bytes</a></div><div class="desc docblock-short">Serialize a message from bytes using bincode.</div></li><li><div class="item-name"><a class="fn" href="fn.ipv4_resolve.html" title="fn hydroflow::util::ipv4_resolve">ipv4_<wbr>resolve</a></div><div class="desc docblock-short">Resolve the <code>ipv4</code> <a href="https://doc.rust-lang.org/nightly/core/net/socket_addr/enum.SocketAddr.html" title="enum core::net::socket_addr::SocketAddr"><code>SocketAddr</code></a> from an IP or hostname string.</div></li><li><div class="item-name"><a class="fn" href="fn.iter_batches_stream.html" title="fn hydroflow::util::iter_batches_stream">iter_<wbr>batches_<wbr>stream</a></div><div class="desc docblock-short">Converts an iterator into a stream that emits <code>n</code> items at a time, yielding between each batch.</div></li><li><div class="item-name"><a class="fn" href="fn.ready_iter.html" title="fn hydroflow::util::ready_iter">ready_<wbr>iter</a></div><div class="desc docblock-short">Returns an <a href="https://doc.rust-lang.org/nightly/core/iter/traits/iterator/trait.Iterator.html" title="trait core::iter::traits::iterator::Iterator"><code>Iterator</code></a> of any immediately available items from the [<code>Stream</code>].</div></li><li><div class="item-name"><a class="fn" href="fn.run_cargo_example.html" title="fn hydroflow::util::run_cargo_example">run_<wbr>cargo_<wbr>example</a></div><div class="desc docblock-short">Run a rust example as a test.</div></li><li><div class="item-name"><a class="fn" href="fn.serialize_to_bytes.html" title="fn hydroflow::util::serialize_to_bytes">serialize_<wbr>to_<wbr>bytes</a></div><div class="desc docblock-short">Serialize a message to bytes using bincode.</div></li><li><div class="item-name"><a class="fn" href="fn.sort_unstable_by_key_hrtb.html" title="fn hydroflow::util::sort_unstable_by_key_hrtb">sort_<wbr>unstable_<wbr>by_<wbr>key_<wbr>hrtb</a></div><div class="desc docblock-short">Sort a slice using a key fn which returns references.</div></li><li><div class="item-name"><a class="fn" href="fn.tcp_bytes.html" title="fn hydroflow::util::tcp_bytes">tcp_<wbr>bytes</a></div><div class="desc docblock-short">Helper creates a TCP <code>Stream</code> and <code>Sink</code> for <code>Bytes</code> strings where each string is
length-delimited.</div></li><li><div class="item-name"><a class="fn" href="fn.tcp_bytestream.html" title="fn hydroflow::util::tcp_bytestream">tcp_<wbr>bytestream</a></div><div class="desc docblock-short">Helper creates a TCP <code>Stream</code> and <code>Sink</code> for undelimited streams of <code>Bytes</code>.</div></li><li><div class="item-name"><a class="fn" href="fn.tcp_framed.html" title="fn hydroflow::util::tcp_framed">tcp_<wbr>framed</a></div><div class="desc docblock-short">Helper creates a TCP <code>Stream</code> and <code>Sink</code> from the given socket, using the given <code>Codec</code> to
handle delineation between inputs/outputs.</div></li><li><div class="item-name"><a class="fn" href="fn.tcp_lines.html" title="fn hydroflow::util::tcp_lines">tcp_<wbr>lines</a></div><div class="desc docblock-short">Helper creates a TCP <code>Stream</code> and <code>Sink</code> for <code>str</code>ings delimited by newlines.</div></li><li><div class="item-name"><a class="fn" href="fn.udp_bytes.html" title="fn hydroflow::util::udp_bytes">udp_<wbr>bytes</a></div><div class="desc docblock-short">Helper creates a UDP <code>Stream</code> and <code>Sink</code> for <code>Bytes</code> strings where each string is
length-delimited.</div></li><li><div class="item-name"><a class="fn" href="fn.udp_bytestream.html" title="fn hydroflow::util::udp_bytestream">udp_<wbr>bytestream</a></div><div class="desc docblock-short">Helper creates a UDP <code>Stream</code> and <code>Sink</code> for undelimited streams of <code>Bytes</code>.</div></li><li><div class="item-name"><a class="fn" href="fn.udp_framed.html" title="fn hydroflow::util::udp_framed">udp_<wbr>framed</a></div><div class="desc docblock-short">Returns a UDP <code>Stream</code>, <code>Sink</code>, and address for the given socket, using the given <code>Codec</code> to
handle delineation between inputs/outputs.</div></li><li><div class="item-name"><a class="fn" href="fn.udp_lines.html" title="fn hydroflow::util::udp_lines">udp_<wbr>lines</a></div><div class="desc docblock-short">Helper creates a UDP <code>Stream</code> and <code>Sink</code> for <code>String</code>s delimited by newlines.</div></li><li><div class="item-name"><a class="fn" href="fn.unbounded_channel.html" title="fn hydroflow::util::unbounded_channel">unbounded_<wbr>channel</a></div><div class="desc docblock-short">Returns a channel as a (1) unbounded sender and (2) unbounded receiver <code>Stream</code> for use in Hydroflow.</div></li><li><div class="item-name"><a class="fn" href="fn.unix_bytes.html" title="fn hydroflow::util::unix_bytes">unix_<wbr>bytes</a></div><div class="desc docblock-short">Helper creates a Unix <code>Stream</code> and <code>Sink</code> for <code>Bytes</code> strings where each string is
length-delimited.</div></li><li><div class="item-name"><a class="fn" href="fn.unix_bytestream.html" title="fn hydroflow::util::unix_bytestream">unix_<wbr>bytestream</a></div><div class="desc docblock-short">Helper creates a Unix <code>Stream</code> and <code>Sink</code> for undelimited streams of <code>Bytes</code>.</div></li><li><div class="item-name"><a class="fn" href="fn.unix_framed.html" title="fn hydroflow::util::unix_framed">unix_<wbr>framed</a></div><div class="desc docblock-short">Helper creates a Unix <code>Stream</code> and <code>Sink</code> from the given socket, using the given <code>Codec</code> to
handle delineation between inputs/outputs.</div></li><li><div class="item-name"><a class="fn" href="fn.unix_lines.html" title="fn hydroflow::util::unix_lines">unix_<wbr>lines</a></div><div class="desc docblock-short">Helper creates a Unix <code>Stream</code> and <code>Sink</code> for <code>str</code>ings delimited by newlines.</div></li><li><div class="item-name"><a class="fn" href="fn.unsync_channel.html" title="fn hydroflow::util::unsync_channel">unsync_<wbr>channel</a></div><div class="desc docblock-short">Returns an unsync channel as a (1) sender and (2) receiver <code>Stream</code> for use in Hydroflow.</div></li><li><div class="item-name"><a class="fn" href="fn.wait_for_process_output.html" title="fn hydroflow::util::wait_for_process_output">wait_<wbr>for_<wbr>process_<wbr>output</a></div><div class="desc docblock-short">Waits for a specific process output before returning.</div></li></ul><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.TcpFramedSink.html" title="type hydroflow::util::TcpFramedSink">TcpFramed<wbr>Sink</a></div><div class="desc docblock-short">A framed TCP <code>Sink</code> (sending).</div></li><li><div class="item-name"><a class="type" href="type.TcpFramedStream.html" title="type hydroflow::util::TcpFramedStream">TcpFramed<wbr>Stream</a></div><div class="desc docblock-short">A framed TCP <code>Stream</code> (receiving).</div></li><li><div class="item-name"><a class="type" href="type.UdpBytesSink.html" title="type hydroflow::util::UdpBytesSink">UdpBytes<wbr>Sink</a></div><div class="desc docblock-short">A UDP undelimited bytes <code>Sink</code> (sending).</div></li><li><div class="item-name"><a class="type" href="type.UdpBytesStream.html" title="type hydroflow::util::UdpBytesStream">UdpBytes<wbr>Stream</a></div><div class="desc docblock-short">A UDP undelimited bytes <code>Stream</code> (receiving).</div></li><li><div class="item-name"><a class="type" href="type.UdpFramedSink.html" title="type hydroflow::util::UdpFramedSink">UdpFramed<wbr>Sink</a></div><div class="desc docblock-short">A framed UDP <code>Sink</code> (sending).</div></li><li><div class="item-name"><a class="type" href="type.UdpFramedStream.html" title="type hydroflow::util::UdpFramedStream">UdpFramed<wbr>Stream</a></div><div class="desc docblock-short">A framed UDP <code>Stream</code> (receiving).</div></li><li><div class="item-name"><a class="type" href="type.UdpLinesSink.html" title="type hydroflow::util::UdpLinesSink">UdpLines<wbr>Sink</a></div><div class="desc docblock-short">A UDP newline-delimited <code>String</code> <code>Sink</code> (sending).</div></li><li><div class="item-name"><a class="type" href="type.UdpLinesStream.html" title="type hydroflow::util::UdpLinesStream">UdpLines<wbr>Stream</a></div><div class="desc docblock-short">A UDP newline-delimited <code>String</code> <code>Stream</code> (receivng).</div></li><li><div class="item-name"><a class="type" href="type.UdpSink.html" title="type hydroflow::util::UdpSink">UdpSink</a></div><div class="desc docblock-short">A UDP length-delimited frame <code>Sink</code> (sending).</div></li><li><div class="item-name"><a class="type" href="type.UdpStream.html" title="type hydroflow::util::UdpStream">UdpStream</a></div><div class="desc docblock-short">A UDP length-delimited frame <code>Stream</code> (receiving).</div></li></ul></section></div></main></body></html>