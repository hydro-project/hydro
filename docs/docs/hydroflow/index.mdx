import {Redirect} from '@docusaurus/router';

<Redirect to="/docs/dfir" />
