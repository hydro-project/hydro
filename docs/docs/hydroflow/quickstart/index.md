# Quickstart
This section will get you up and running with Rust and Hydroflow, and work you through a set of concrete examples illustrating much of what you need to know to get productive with Hydroflow.
