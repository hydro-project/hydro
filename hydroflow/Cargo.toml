[package]
name = "hydroflow"
version = "0.1.0"
edition = "2021"

[features]
default = [ "async", "macros" ]
async = [ "futures" ]
macros = [ "hydroflow_macro", "hydroflow_datalog" ]

[[example]]
name = "echoserver"

[[example]]
name = "echo_serde_json"

[[example]]
name = "chat"

[[example]]
name = "graph_reachability"

[[example]]
name = "graph_unreachability"

[[example]]
name = "three_clique"

[[example]]
name = "two_pc"

[[example]]
name = "deadlock_detector"

[[example]]
name = "kvs"

[dependencies]
bincode = "1.3"
byteorder = "1.4.3"
bytes = "1.1.0"
futures = { version = "0.3", optional = true }
hydroflow_datalog = { optional = true, path = "../hydroflow_datalog" }
hydroflow_lang = { path = "../hydroflow_lang" }
hydroflow_macro = { optional = true, path = "../hydroflow_macro" }
pin-project-lite = "0.2.9"
pusherator = { path = "../pusherator" }
ref-cast = "1.0"
regex = "1"
sealed = "0.4"
serde = { version = "1", features = [ "derive" ] }
serde_json = "1"
slotmap = "1.0"
static_assertions = "1.1.0"
tokio = { version = "1.16", features = [ "full" ] }
tokio-stream = { version = "0.1.10", features = [ "io-util" ] }
tokio-util = { version = "0.7.4", features = [ "net", "codec" ] }
variadics = { path = "../variadics" }

[dev-dependencies]
chrono = { version = "0.4.20", features = [ "serde" ], default-features = false }
clap = { version = "3.0.14", features = [ "derive" ] }
colored = "2.0"
criterion = { version = "0.3", features = [ "async_tokio" ] }
futures = { version = "0.3" }
itertools = "0.10.3"
rand = "0.8.4"
textnonce = "1.0.0"
time = "0.3"
trybuild = "1.0.71"
hdrhistogram = "7"
zipf = "7.0.0"
core_affinity = "0.5.10"
