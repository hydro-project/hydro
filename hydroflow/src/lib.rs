#![feature(never_type)]
#![feature(type_alias_impl_trait)]
#![allow(type_alias_bounds)]
#![allow(clippy::let_and_return)]
#![allow(clippy::iter_with_drain)]
#![allow(clippy::explicit_auto_deref)]

pub mod builder;
pub mod compiled;
pub mod lang;
pub mod props;
pub mod scheduled;

pub use hydroflow_macro::*;

pub use pusherator;
pub use tokio;
pub use tuple_list::tuple_list as tl;
pub use tuple_list::tuple_list_type as tt;

#[cfg(doctest)]
mod booktest {
    macro_rules! booktest {
        ($i:ident) => {
            #[doc = include_str!(concat!("../../book/", stringify!($i), ".md"))]
            mod $i {}
        };
    }
    booktest!(example_1);
    booktest!(example_2);
    booktest!(example_3);
    booktest!(example_4);
    booktest!(example_5);

    booktest!(example_1_surface);
    booktest!(example_2_surface);
    booktest!(example_3_surface);
    booktest!(example_4_surface);
    booktest!(example_5_surface);

    booktest!(surface_syntax);
    booktest!(surface_embedding);
    booktest!(surface_flows);
    booktest!(surface_data);
    booktest!(surface_ops);
}
