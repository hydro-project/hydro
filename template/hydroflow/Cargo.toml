[package]
name = "hydroflow-template"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
chrono = { version = "0.4.20", features = [ "serde" ], default-features = true }
clap = { version = "4.0.29", features = [ "derive" ] }
hydroflow = { git = "{{ hydroflow_git | default: 'https://github.com/hydro-project/hydroflow.git' }}", branch = "{{ hydroflow_branch | default: 'main' }}" }
serde = { version = "1", features = [ "derive" ] }
