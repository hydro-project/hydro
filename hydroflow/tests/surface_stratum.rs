use hydroflow::hydroflow_syntax;

// RUSTFLAGS="$RUSTFLAGS -Z proc-macro-backtrace" cargo test --package hydroflow --test surface_stratum -- test_surface_syntax_strata --exact --nocapture
#[test]
pub fn test_surface_syntax_strata() {
    // An edge in the input data = a pair of `usize` vertex IDs.
    let (pairs_send, pairs_recv) = tokio::sync::mpsc::unbounded_channel::<(usize, usize)>();

    // let mut df = hydroflow_syntax! {
    //     reached_vertices = merge() -> map(|v| (v, ()));
    //     recv_iter(vec![0]) -> [0]reached_vertices;

    //     my_join_tee = join() -> map(|(_src, ((), dst))| dst) -> map(|x| x) -> next_stratum() -> map(|x| x) -> tee();
    //     reached_vertices -> [0]my_join_tee;
    //     recv_stream(pairs_recv) -> [1]my_join_tee;

    //     my_join_tee[0] -> [1]reached_vertices;
    //     my_join_tee[1] -> next_stratum() -> for_each(|x| println!("Reached: {}", x));
    // };
    let mut df = hydroflow_syntax! {
        reached_vertices = merge() -> map(|v| (v, ()));
        recv_iter(vec![0]) -> [0]reached_vertices;

        my_join_tee = join() -> map(|(_src, ((), dst))| dst) -> map(|x| x) -> map(|x| x) -> tee();
        reached_vertices -> [0]my_join_tee;
        recv_stream(pairs_recv) -> [1]my_join_tee;

        my_join_tee[0] -> [1]reached_vertices;
        my_join_tee[1] -> for_each(|x| println!("Reached: {}", x));
    };

    println!(
        "{}",
        df.serde_graph()
            .expect("No graph found, maybe failed to parse.")
            .to_mermaid()
    );
    df.run_available();

    pairs_send.send((0, 1)).unwrap();
    df.run_available();

    pairs_send.send((2, 4)).unwrap();
    pairs_send.send((3, 4)).unwrap();
    df.run_available();

    pairs_send.send((1, 2)).unwrap();
    df.run_available();

    pairs_send.send((0, 3)).unwrap();
    df.run_available();

    pairs_send.send((0, 3)).unwrap();
    df.run_available();

    // Reached: 1
    // Reached: 2
    // Reached: 4
    // Reached: 3
    // Reached: 4
}
