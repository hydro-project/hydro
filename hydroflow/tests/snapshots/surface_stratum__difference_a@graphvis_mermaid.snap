---
source: hydroflow/tests/surface_stratum.rs
expression: df.meta_graph().unwrap().to_mermaid()
---
%%{init:{'theme':'base','themeVariables':{'clusterBkg':'#ddd','clusterBorder':'#888'}}}%%
flowchart TD
classDef pullClass fill:#8af,stroke:#000,text-align:left,white-space:pre
classDef pushClass fill:#ff8,stroke:#000,text-align:left,white-space:pre
linkStyle default stroke:#aaa,stroke-width:4px,color:red,font-size:1.5em;
subgraph sg_1v1 ["sg_1v1 stratum 0"]
    3v1[\"(3v1) <code>source_iter([1, 3, 5, 7])</code>"/]:::pullClass
end
subgraph sg_2v1 ["sg_2v1 stratum 1"]
    2v1[\"(2v1) <code>source_iter([1, 2, 3, 4])</code>"/]:::pullClass
    1v1[\"(1v1) <code>difference()</code>"/]:::pullClass
    4v1[/"(4v1) <code>for_each(|x| output_inner.borrow_mut().insert(x))</code>"\]:::pushClass
    2v1--pos--->1v1
    1v1--->4v1
    subgraph sg_2v1_var_a ["var <tt>a</tt>"]
        1v1
    end
end
3v1--->5v1
5v1["(5v1) <code>handoff</code>"]:::otherClass
5v1==neg===o1v1

